File: api/src/test/java/net/blueshell/api/factory/dto/survey/AnswerDTOFactory.java
package net.blueshell.api.factory.dto.survey;

import lombok.RequiredArgsConstructor;
import net.blueshell.api.dto.survey.AnswerDTO;
import net.blueshell.api.dto.survey.QuestionDTO;
import net.blueshell.api.factory.dto.BaseDtoFactory;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

/**
 * Factory for AnswerDTO test instances.
 */
@Component
@RequiredArgsConstructor
public class AnswerDTOFactory extends BaseDtoFactory<AnswerDTO> {

    @Override
    public Class<AnswerDTO> targetType() {
        return AnswerDTO.class;
    }

    @Override
    public AnswerDTO createBasic() {
        AnswerDTO dto = new AnswerDTO();
        dto.setQuestionId(nextId());
        dto.setTextResponse("Sample text response");
        dto.setOptionSelections(null);
        return dto;
    }

    public AnswerDTO createForQuestion(QuestionDTO question) {
        AnswerDTO dto = new AnswerDTO();
        dto.setQuestionId(question.getId() != null ? question.getId() : nextId());

        switch (question.getType()) {
            case OPEN -> {
                dto.setTextResponse("This is a text response for the open question");
                dto.setOptionSelections(null);
            }
            case RADIO -> {
                dto.setTextResponse(null);
                if (question.getChoiceLabels() != null && !question.getChoiceLabels().isEmpty()) {
                    List<Boolean> selections = new ArrayList<>();
                    for (int i = 0; i < question.getChoiceLabels().size(); i++) {
                        selections.add(i == 0);
                    }
                    dto.setOptionSelections(selections);
                }
            }
            case CHECKBOX -> {
                dto.setTextResponse(null);
                if (question.getChoiceLabels() != null && !question.getChoiceLabels().isEmpty()) {
                    List<Boolean> selections = new ArrayList<>();
                    for (int i = 0; i < question.getChoiceLabels().size(); i++) {
                        selections.add(random.nextBoolean());
                    }
                    dto.setOptionSelections(selections);
                }
            }
            case DESCRIPTION -> {
                dto.setTextResponse(null);
                dto.setOptionSelections(null);
            }
            default -> {
                dto.setTextResponse("Default response");
                dto.setOptionSelections(null);
            }
        }
        return dto;
    }

    public AnswerDTO createForOpenQuestion() {
        AnswerDTO dto = new AnswerDTO();
        dto.setQuestionId(nextId());
        dto.setTextResponse("Detailed response to the open-ended question");
        dto.setOptionSelections(null);
        return dto;
    }

    public AnswerDTO createForRadioQuestion(int optionCount, int selectedIndex) {
        AnswerDTO dto = new AnswerDTO();
        dto.setQuestionId(nextId());
        dto.setTextResponse(null);

        List<Boolean> selections = new ArrayList<>();
        for (int i = 0; i < optionCount; i++) {
            selections.add(i == selectedIndex);
        }
        dto.setOptionSelections(selections);
        return dto;
    }

    public AnswerDTO createForCheckboxQuestion(int optionCount, List<Integer> selectedIndices) {
        AnswerDTO dto = new AnswerDTO();
        dto.setQuestionId(nextId());
        dto.setTextResponse(null);

        List<Boolean> selections = new ArrayList<>();
        for (int i = 0; i < optionCount; i++) {
            selections.add(selectedIndices.contains(i));
        }
        dto.setOptionSelections(selections);
        return dto;
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/dto/survey/AnswerDTOFactory.kt
package net.blueshell.api.factory.dto.survey

import net.blueshell.api.shared.enums.QuestionType
import net.blueshell.api.survey.api.dto.AnswerDTO
import net.blueshell.api.survey.api.dto.QuestionDTO
import net.blueshell.api.survey.api.dto.SurveyDTO
import net.blueshell.api.factory.dto.BaseDtoFactory
import org.springframework.stereotype.Component

/**
 * Factory for AnswerDTO test instances.
 */
@Component
class AnswerDTOFactory : BaseDtoFactory<AnswerDTO>() {

    override fun targetType(): Class<AnswerDTO> = AnswerDTO::class.java

    override fun createBasic(): AnswerDTO {
        val dto = AnswerDTO()
        dto.questionId = nextId()
        dto.textResponse = "Sample text response"
        dto.optionSelections = null
        return dto
    }

    fun createForQuestion(question: QuestionDTO): AnswerDTO {
        val dto = AnswerDTO()
        dto.questionId = question.id ?: nextId()

        when (question.type) {
            QuestionType.OPEN -> {
                dto.textResponse = "This is a text response for the open question"
                dto.optionSelections = null
            }
            QuestionType.RADIO -> {
                dto.textResponse = null
                val choices = question.choiceLabels
                if (!choices.isNullOrEmpty()) {
                    val selections = MutableList(choices.size) { index -> index == 0 }
                    dto.optionSelections = selections
                }
            }
            QuestionType.CHECKBOX -> {
                dto.textResponse = null
                val choices = question.choiceLabels
                if (!choices.isNullOrEmpty()) {
                    val selections = MutableList(choices.size) { random.nextBoolean() }
                    dto.optionSelections = selections
                }
            }
            QuestionType.DESCRIPTION -> {
                dto.textResponse = null
                dto.optionSelections = null
            }
            else -> {
                dto.textResponse = "Default response"
                dto.optionSelections = null
            }
        }
        return dto
    }

    fun createForOpenQuestion(): AnswerDTO {
        val dto = AnswerDTO()
        dto.questionId = nextId()
        dto.textResponse = "Detailed response to the open-ended question"
        dto.optionSelections = null
        return dto
    }

    fun createForRadioQuestion(optionCount: Int, selectedIndex: Int): AnswerDTO {
        val dto = AnswerDTO()
        dto.questionId = nextId()
        dto.textResponse = null

        val selections = MutableList(optionCount) { index -> index == selectedIndex }
        dto.optionSelections = selections
        return dto
    }

    fun createForCheckboxQuestion(optionCount: Int, selectedIndices: List<Int>): AnswerDTO {
        val dto = AnswerDTO()
        dto.questionId = nextId()
        dto.textResponse = null

        val selections = MutableList(optionCount) { index -> selectedIndices.contains(index) }
        dto.optionSelections = selections
        return dto
    }

    fun createForSurvey(survey: SurveyDTO): MutableList<AnswerDTO> {
        return survey.questions.map { question -> createForQuestion(question) }.toMutableList()
    }
}
